src/rebase.js

```diff
@@ -116,13 +116,22 @@ module.exports = (function(){
   function _fetch(endpoint, options){
     _validateEndpoint(endpoint);
     optionValidators.context(options);
-    optionValidators.then(options);
     options.queries && optionValidators.query(options);
     var ref = firebase.database().ref(endpoint);
     ref = _addQueries(ref, options.queries);
-    ref.once('value', (snapshot) => {
+    return ref.once('value').then(snapshot => {
       var data = options.asArray === true ? _toArray(snapshot) : snapshot.val();
-      options.then.call(options.context, data);
+      if(options.then){
+        options.then.call(options.context, data);
+      }
+      return data;
+    }, err => {
+      //call onFailure callback if it exists otherwise return a rejected promise
+      if(options.onFailure && typeof options.onFailure === 'function'){
+        options.onFailure.call(options.context, err);
+      } else {
```

Comment:
@anotherpit good question on the other methods. I sure hope not, but I will definitely double check tonight. 


> Second, if we use ref.once('value', onSuccess, onFailure, context) instead of ref.once('value').then(...) here, will it produce such an error on failure?

That is what I was thinking of trying in the comment on line 123 below. Not calling `.then` for the user. Instead use just the onSuccess and onFailure callbacks and return a Promise they can call `.then` on and assume responsibility for adding `.catch`